### PR TITLE
Feature/andrew yash/frontend contexts

### LINKED ISSUE #15: User Authentication: Frontend Contexts
# Overview

### Auth Context

We will also need a way to track the global state of authentication throughout the app. This will be helpful later on for the profile page as well as for any page where we require a user to be logged in to access. We can do this using a React Context.

Create an Auth Context that tracks two values: `currentUser` (should be set to the Firebase user object on successful sign-in and should be `null` when the user is signed out) and `signedIn` (`true` if the user is signed in, `false` otherwise). See [this example](https://github.com/TritonSE/TAP-Foster-Dog-Matching/blob/main/client/src/contexts/AuthContext.js). Use [`onAuthStateChanged`](https://firebase.google.com/docs/auth/web/manage-users#get_the_currently_signed-in_user) to get the Firebase user object. NOTE: The handler you provide is called whenever the auth state changes (e.g. when a user logs in or logs out). Wrap the application with the provider for the context you created like `ThemeProvider` in `App.tsx`.

### Data Context

We also need a Context to store data that should be accessible throughout the app. In a way, this Context will act as a cache for data that should not change frequently during a session. This reduces the number of redundant network calls we need to make to the backend, thus keeping the app as performant as possible for our users.

First, create a `DataContext` that will cache global app data ([example](https://github.com/TritonSE/TAP-Foster-Dog-Matching/blob/main/client/src/contexts/DataContext.js)). 

One example of data we want to cache is the list of users (i.e. HLs and CMs). This data is used throughout the application (such as in the dropdowns for selecting the assigned HL and CM for referrals as well as the upcoming Roster page). Additionally, this data is not expected to change a lot during the lifetime of a session. To cache users, `DataContext` should:

1. Call `GET /api/users` to fetch the list of all users (using `getUsers` function in `frontend/src/api/users.ts`)
2. Provide an object with these values:
    - `allHousingLocators`: array of all HLs (helpful for HL dropdown)
    - `allCaseManagers`: array of all CMs (helpful for CM dropdown)
    - `currentUser`: the User object matching the email address of the current Firebase user (helpful for Profile page later)

NOTE: the `AuthContext`'s `currentUser` is the Firebase user object while the `DataContext`'s `currentUser` is the User object that is stored in our database. Although both objects have some overlapping info about the user, we still need both objects because the Firebase user object has the user's profile image link and the User object tells us whether the user is a HL or not.

Wrap the application with the provider for the context you created like `AuthContext` in `App.tsx`.

## Acceptance Criteria
- [x] Auth Context is implemented and the provider is used in `App.tsx`
- [x] Data Context is implemented and the provider is used in `App.tsx`

### PR DESCRIPTION
**Pull request by YashRavipati1**
## Tracking Info

Resolves #15 

## Changes

Adding auth and data contexts to use instead of making backend/firebase calls every time we need user information.

## Testing

Tested the contexts on the Login page by printing out user information from the contexts.

## Confirmation of Change

Screenshot 2024-02-05 at 6 17 58 PM

### CODE REVIEW

File: frontend/src/pages/Login.tsx [hunk 1]
```diff
@@ -1,14 +1,110 @@
...
+
+  const authContext = useContext(AuthContext);
+  const dataContext = useContext(DataContext);
+  const navigate = useNavigate();
+
+  useEffect(() => {
+    console.log(authContext.currentUser);
+    console.log(authContext.signedIn);
+    console.log(dataContext.currentUser);
+  });
```
Comment (petabite) [inline frontend/src/pages/Login.tsx]: Do we need this anymore?

File: frontend/src/pages/Home.tsx [hunk 2]
```diff
@@ -1,21 +1,54 @@
...
+      .catch((error) => {
+        console.log(error);
+      });
+  };
+
+  const getInfo = () => {
+    console.log(authContext.currentUser);
+    console.log("Signed in?: " + authContext.signedIn);
+    console.log(dataContext.currentUser);
+  };
```
Comment (petabite) [inline frontend/src/pages/Home.tsx]: This too?

File: frontend/src/contexts/AuthContext.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+  useEffect(() => {
+    onAuthStateChanged(auth, (user: User | null) => {
+      if (user) {
+        setUser(user);
+      } else {
+        setUser(null);
+      }
+    });
+  });
```
Comment (petabite) [inline frontend/src/contexts/AuthContext.tsx:28]: Just as good practice, it is a good idea to [return a cleanup function](https://blog.logrocket.com/understanding-react-useeffect-cleanup-function/) ([see example](https://github.com/TritonSE/TAP-Foster-Dog-Matching/blob/101fb04c11dd45d90e6956a7c417414006bbafe4/client/src/contexts/AuthContext.js#L50-L70)) so that the listener is removed after the component is destroyed. Doing this prevents dangling listeners (and keeps the app free of memory-leaks). Also, currently, the code is registering a new listener every time the component rerenders (we only want a listener to get registered once). Passing `[]` as the dependencies for the effect will ensure the effect runs only once (when the component renders for the first time).

```suggestion
  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user: User | null) => {
      if (user) {
        setUser(user);
      } else {
        setUser(null);
      }
    });
    return unsubscribe;
  }, []);
```

File: frontend/src/contexts/DataContext.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+export function DataProvider({ children }: ProviderProps) {
+  const [allHousingLocators, setAllHousingLocators] = useState<User[]>([]);
+  const [allCaseManagers, setAllCaseManagers] = useState<User[]>([]);
+  const [currentUser, setUser] = useState<User | null>(null);
+  const auth = useContext(AuthContext);
+
+  const fetchData = useCallback(async () => {
+    const jsonData = await get("/users");
+    const data = (await jsonData.json()) as User[];
```
Comment (petabite) [inline frontend/src/contexts/DataContext.tsx]: Maybe use the `getUsers()` function in `frontend/src/api/users.ts`?

**Review by petabite — COMMENTED** [review]
Looks good! Left some comments. After those, rebase this branch on main (and resolve the merge conflicts) and you should be good to go

File: frontend/src/contexts/AuthContext.tsx (hunk 3 above)
Comment (Azhou2023) [reply frontend/src/contexts/AuthContext.tsx:28]: Fixed all of this in my NavBar pr, forgot to change it here too, sorry about that. Would it be fine to leave this as is and have it be fixed by the nav pr? (I made a lot of other changes to the file in that pr too)

Comment (petabite) [reply frontend/src/contexts/AuthContext.tsx:28]: Oh gotcha. Just saw that lol. Yep, that works. Feel free to ignore this comment then.

File: frontend/src/pages/Login.tsx (hunk 1 above)
Comment (YashRavipati1) [reply frontend/src/pages/Login.tsx]: Removed

File: frontend/src/pages/Home.tsx (hunk 2 above)
Comment (YashRavipati1) [reply frontend/src/pages/Home.tsx]: Removed

File: frontend/src/contexts/DataContext.tsx (hunk 4 above)
Comment (YashRavipati1) [reply frontend/src/contexts/DataContext.tsx]: Changed to use frontend API functions

**Review by JamesZhang878 — APPROVED** [review]
LGTM! When resolving merge conflicts with main, please be sure to use the version of backend/src/firebase-admin.ts from main because my code for the middleware won't work otherwise. Thanks!

File: backend/src/server.ts [hunk 5]
```diff
@@ -33,7 +33,7 @@ export const backend = onRequest({ region: "us-west1" }, app);
 
 export const beforecreated = beforeUserCreated((event) => {
   const user = event.data;
-  if (ENVIRONMENT === "production" && !user.email?.includes("unionstationhs.org")) {
+  if (ENVIRONMENT === "staging" && !user.email?.includes("@ucsd.edu")) {
```
Comment (petabite) [inline backend/src/server.ts]: Can we revert this change?

```suggestion
  if (ENVIRONMENT === "production" && !user.email?.includes("unionstationhs.org")) {
```

File: frontend/src/pages/Login.tsx [hunk 6]
```diff
@@ -78,7 +77,7 @@ export function Login() {
       })
       .catch((error: FirebaseError) => {
         console.log(error);
-        if (error.code === "auth/internal-error" && error.message.includes("Cloud Function")) {
+        if (error.code !== "auth/internal-error" && error.message.includes("Cloud Function")) {
```
Comment (petabite) [inline frontend/src/pages/Login.tsx]: This too
```suggestion
        if (error.code === "auth/internal-error" && error.message.includes("Cloud Function")) {
```

**Review by petabite — CHANGES_REQUESTED** [review]

File: backend/src/server.ts (hunk 5 above)
Comment (YashRavipati1) [reply backend/src/server.ts]: Reverted

File: frontend/src/pages/Login.tsx (hunk 6 above)
Comment (YashRavipati1) [reply frontend/src/pages/Login.tsx]: Reverted

**Review by petabite — APPROVED** [review]
LGTM 🔥 Thanks for addressing the comments! 

I removed the AuthContext testing to prevent merge conflicts later and fixed some tiny things.
